docs/designs/checkpoint_design.md

```diff
@@ -0,0 +1,59 @@
+# ElasticDL Training Checkpoints Design
+This document describes the design of save checkpoints for ElasticDL training.
+
+## Motivation
+Checkpoints capture the exact value of all parameters (tf.Variable objects and embedding vectors) used by a model. In ElasticDL, we need to save all parameters of the model to checkpoints during training for the following reasons:
+* We can export the model with the best evaluation metrics during training from checkpoints. Generally, we evaluate the model and monitor the evaluation metrics during training. We will stop training when we find the evaluation metrics don't go better or even worse due to over-fitting. So, the final parameters may not be the best. With checkpoints, we can choose a checkpoint with the best evaluation metrics to export for serving.
+* We can restore the model parameters from a checkpoint to train or predict. Fault tolerance in ElasticDL can not resolve all unexpected error e.g. a power outage and OS fault in the master. We are going to have a bad time if we lose one or more of our experiments due to those errors. Other times, even if we don't experience an unforeseen error, we might just want to resume a particular state of the training for a new experiment or try different things from a given state.
+
+In the following sections, we will describe the design of how to make a checkpoint and restore model parameters from a checkpoint in ElasticDL.
+
+## Design Components
+This design contains two key parts: export and restore.
+
+### Export Model Parameters to a Checkpoint
+
+**When to save a checkpoint during training.** 
+
+In ElasticDL, we will save all model parameters to a checkpoint directory every `checkpoint_steps`. Though, some frameworks support to save checkpoints every N steps or every N seconds, e.g. [Estimator](https://www.tensorflow.org/api_docs/python/tf/estimator/RunConfig#__init__). Using ParameterServerStrategy in ElasticDL, Communication with PS is asynchronous and we cannot guarantee that the iteration steps of all PS instances are consistent at the same time. So, we don't support to save checkpoints every N seconds.
```

Comment:
Step and time are similar concepts. Why time is not supported?

Reply:
At the same, the iteration steps of all PS instances may be different, because the iteration speed of different PS instances is different due to different CPU, IO and so on.

Reply:
`the iteration steps` -> `iteration steps`